features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+                                               const uint8_t * val)
+{
+  /* Check data is 32bit aligned, if not, read into temporary buffer and
+     then move to user buffer. */
+  if ((uint32_t)val & 0x3)
+  {
+    uint32_t temp[4];
+    CRYPTO_DataRead(reg, temp);
+    memcpy((void*)val, temp, 16);
+  }
+  else
+  {
+    CRYPTO_DataRead(reg, (uint32_t*)val);
+  }
+}
+
+__STATIC_INLINE void CRYPTO_DataWriteUnaligned(volatile uint32_t * reg,
+                                               const uint8_t * val)
+{
+  /* Check data is 32bit aligned, if not move to temporary buffer before
+     writing.*/
+  if ((uint32_t)val & 0x3)
+  {
+    uint32_t temp[4];
+    memcpy(temp, val, 16);
+    CRYPTO_DataWrite(reg, temp);
+  }
+  else
+  {
+    CRYPTO_DataWrite(reg, (uint32_t*)val);
+  }
+}
+
+/*
+ * Initialize AES context
+ */
+void mbedtls_aes_init( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/*
+ * Clear AES context
+ */
+void mbedtls_aes_free( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/*
+ * AES key schedule (encryption)
+ */
+int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx,
+                            const unsigned char *key,
+                            unsigned int keybits )
+{
+    if( ctx == NULL || key == NULL ) {
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+
+    if ( ( 128UL != keybits ) && ( 256UL != keybits ) ) {
+        /* Unsupported key size */
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    ctx->keybits = keybits;
+    memcpy(ctx->key, key, keybits/8);
+
+    return 0;
+}
+
+/*
+ * AES key schedule (decryption)
+ */
+int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx,
+                            const unsigned char *key,
+                            unsigned int keybits )
+{
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || key == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+
+    if ( ( 128UL != keybits ) && ( 256UL != keybits ) ) {
+        /* Unsupported key size */
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    ctx->keybits = keybits;
+
+    CRYPTO_TypeDef *device = crypto_management_acquire();
+    device->WAC = 0;
+    device->CTRL = 0;
+
+    CORE_ENTER_CRITICAL();
+    CRYPTO_KeyBufWrite(device, (uint32_t*)key, (keybits == 128) ? cryptoKey128Bits : cryptoKey256Bits);
+    CORE_EXIT_CRITICAL();
+
+    /* Busy-wait here to allow context-switching to occur */
+    device->CMD = CRYPTO_CMD_INSTR_AESENC;
+    while ((device->STATUS & CRYPTO_STATUS_INSTRRUNNING) != 0);
+
+    CORE_ENTER_CRITICAL();
+    CRYPTO_KeyRead(device, (uint32_t*)ctx->key, (keybits == 128) ? cryptoKey128Bits : cryptoKey256Bits);
```

Comment:
You can't do a memcopy since the key is located in a 128-bit register which is only accessible through 4 consecutive writes/reads to/from the same 32-bit gateway register